physical/s3.go

```diff
@@ -0,0 +1,207 @@
+package physical
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/awslabs/aws-sdk-go/aws"
+	"github.com/awslabs/aws-sdk-go/aws/credentials"
+	"github.com/awslabs/aws-sdk-go/service/s3"
+)
+
+// S3Backend is a physical backend that stores data
+// within an S3 bucket.
+type S3Backend struct {
+	bucket string
+	client *s3.S3
+	l      sync.Mutex
+}
+
+// newS3Backend constructs a S3 backend using a pre-existing
+// bucket. Credentials can be provided to the backend, sourced
+// from the environment, AWS credential files or by IAM role.
+func newS3Backend(conf map[string]string) (Backend, error) {
+
+	bucket, ok := conf["bucket"]
+	if !ok {
+		return nil, fmt.Errorf("'bucket' must be set")
+	}
+
+	access_key, ok := conf["access_key"]
+	if !ok {
+		access_key = ""
+	}
+	secret_key, ok := conf["secret_key"]
+	if !ok {
+		secret_key = ""
+	}
+	region, ok := conf["region"]
+	if !ok {
+		region = os.Getenv("AWS_DEFAULT_REGION")
+		if region == "" {
+			region = "us-east-1"
+		}
+	}
+
+	creds := credentials.NewChainCredentials([]credentials.Provider{
+		&credentials.StaticProvider{Value: credentials.Value{
+			AccessKeyID:     access_key,
+			SecretAccessKey: secret_key,
+		}},
+		&credentials.EnvProvider{},
+		&credentials.SharedCredentialsProvider{Filename: "", Profile: ""},
+		&credentials.EC2RoleProvider{},
+	})
+
+	s3conn := s3.New(&aws.Config{
+		Credentials: creds,
+		Region:      region,
+	})
+
+	_, err := s3conn.HeadBucket(&s3.HeadBucketInput{Bucket: &bucket})
+	if err != nil {
+		return nil, fmt.Errorf("unable to access bucket '%s': %v", bucket, err)
+	}
+
+	s := &S3Backend{
+		client: s3conn,
+		bucket: bucket,
+	}
+	return s, nil
+}
+
+// Put is used to insert or update an entry
+func (s *S3Backend) Put(entry *Entry) error {
+	s.l.Lock()
```

Comment:
This internal locking is not necessary, there is no need to serialize the access to S3.